Rework `BufferedRandomAccessFile`

Calvin-L opened the pull request:
This commit makes major alterations to `BufferedRandomAccessFile` to bring it in line with the general `RandomAccessFile` contract:

 - Invariant V1 has been strengthened to include a contract for the `length` field.
 - Invariants V2-V5 have been altered to account for cases where `curr > length`, which the `RandomAccessFile` contract allows.
 - Additionally, invariant V5 has been adjusted to clarify that `dirty=true` does not imply that there are unflushed bytes.  The implication is only one-way: if there are unflushed bytes, then `dirty=true`.
 - The `hi` and `maxHi` fields have been removed.  These largely duplicate information that can be quickly recomputed, meaning they add additional complexity to the class invariants.  Without those fields, invariant V6 can be removed.
 - The `close()` method now closes the underlying file descriptor even if `flush()` throws an exception.

Due to the complexity of this change, I have added a TLA+ specification for `BufferedRandomAccessFile` to enable model checking of its design.

To ensure that the code is correct, this commit also adds a substantial amount of test infrastructure:

 - Several new regression tests have been added.  These are a little cryptic because they were randomly generated, but they do reflect true behaviors of the `RandomAccessFile` contract.
 - A fuzz test has been added.  This test ensures that many sequences of random operations behave correctly.  It also minimizes failures so they can be easily converted to regression tests.

This commit also makes some whitespace and formatting adjustments.

Fixes #835.  I have verified that the tests now pass under Java 21.

lemmy commented:
Impressive!

ahelwer commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+    /\ length = ArrayLen(LogicalFileContent)
+
+\* Inv2 is a bit special.  Most methods restore it just before they return.  It
+\* is generally restored by calling `seeek(curr)`.  But, that behavior is
```
spelling

lemmy replied:
Next to `seek`, `seeek` is a method in BRAF.

Calvin-L replied:
Nope, `seeek` is actually the correct name! Maybe I should add [sic] here.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+\* Copyright (c) 2024, Oracle and/or its affiliates.
```
IANAL: The copyright remains with Oracle, but the DCO makes sure that the TLA+ foundation may use the code in any way it likes within the bound of the open source license (MIT in our case).

ahelwer replied:
I think it would be worth emailing your contact with the Linux Foundation with this question.

Calvin-L replied:
My company requires these notices on any code I write on company hardware and time. (Even for internal code!)

----

If you want more info than that...

I am not a lawyer, so I can't claim to fully understand the requirement, but here's what I know (or think I know anyway):
 - These notices do not change the project _license_ (MIT) in any way. Everyone is still free "to deal in the Software without restriction" and so forth. As you know, Oracle even signed a CLA attesting that it understands that fact.
 - These notices do not claim _exclusive_ copyright, and are compatible with overlapping copyrights from other organizations. Again, see the project license.
 - These notices primarily serve as glorified authorship information, and they might be best read by non-lawyers as meaning "this file contains code that was originally owned by Oracle and that Oracle is therefore legally allowed to release to the world under the terms of this project's license".

On a personal note, I find the notices aesthetically ugly. But, I do plan to follow the letter of the law so my company will let me keep contributing to TLA+ with company resources.

If the notices are unacceptable in new code, I can have a chat with my company's legal team and see if they'll accept something different. I know in the past they've been open to a [simple blanket copyright in the license file](https://github.com/wilcoxjay/mypyvy/blob/3a055dfad3d13fb35b6125ab6f43fa6ea4493b5f/LICENSE), but that could take a while to approve (and it might not ever be approved for a larger project like TLA+).

lemmy replied:
I've been adding a similar Microsoft copyright notice in my contributions, and Microsoft also signed the CLA.  That's how it is done.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+
+===============================================================================
+
+--------------------------- MODULE RandomAccessFile ---------------------------
```
I think it's still worth a brief documentation string here, as it's the place I jumped to first to ensure I understood what was going on.  For many folks, even interacting with files in this manner (rather than wholesale reads or writes) is still quite a low level operation.  A suggestion might be:

> Here we model abstractly how one interacts with files assuming a single thread that reads and writes, given a pointer into the file.  The pointer can be moved to an arbitrary position, or it can read or write data linearly from its current position, which simultaneously advances the pointer.  We can modify the length of the file and if the pointer would fall outside of the bounds after truncation, it is moved to the end of the file.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+
+CONSTANTS
+    Symbols,
+    ArbitrarySymbol,
```
Seems like the purpose and value of `ArbitrarySymbol` is worth calling out (maybe here or in `TruncateOrExtendFile` or there might be some better place).  If I understand it correctly, it seems that its to reign in state space explosion by mixing in a bit of symbolic checking?  It appears as though on could consider all possible extensions of a file using `Symbols`, but the downside of this is that we're going to potentially explore many states.  Whether I understood correctly or not, it's probably worth explaining.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+ArraySlice(a, startInclusive, endExclusive) == [elems |-> SubSeq(a.elems, startInclusive + 1, endExclusive)]
+ArrayConcat(a1, a2) == [elems |-> a1.elems \o a2.elems]
+
+WriteToFile(file, offset, data_to_write) ==
```
It is interesting to me that most of these functions are about Arrays.  Even `WriteToFile` and `ExtendFile` are relatively easy make generic array operations, maybe `OverlapAt` and `TruncateOrExtendWith`, where both take a placeholder as an argument.  It may simply be worth making this an `Array` module, and then `Common` stays more focused on the spec itself `WriteToFile(file, offset, data_to_write) == Array.Overlap(file, data_to_write, offset, ArbitarySymbol)` (I have no idea why I felt compelled to reverse arguments).

I suppose, otherwise I was going to suggest that we add brief documentation on what the semantics of `WriteToFile` and `ExtendFile` were, but in many ways something like `ExtendFile == Array.TurncateOrExtendWith(file, new_length, ArbitrarySymbol)` does this for you.

IamfromSpace replied:
I am probably overthinking these utility functions, but it struck me that maybe the most intuitive implementation involves slice & concat.

```
TruncateOrExtend(a, T, len) == IF len > ArrayLen(a) THEN ArrayConcat(a, Array(T, len - ArrayLen(a))) ELSE ArraySlice(a, 0, len)

OverlayAt(a1, a2, T, i) == ArrayConcat(TruncateOrExtend(a2, T, i), a2)
```

lemmy replied:
Should any of those utility operators be contributed as an Array module to our CommunityModules?

Calvin-L replied:
I added some simple documentation to these operators.

> but it struck me that maybe the most intuitive implementation involves slice & concat

I agree that the current definitions are a little ugly. I wrote them after one or two failed attempts to come up with simpler definitions.

The ones you've given are certainly cleaner than my early attempts, and I will adopt your definition for `TruncateOrExtend`. But I might keep the definition I wrote for `WriteToFile`. The equivalent definition in terms of other operators is actually fairly cryptic (your definition for `OverlayAt` needs to include the portion of the contents after the written data if the contents were not written at the end of the array):

```tla
WriteToFile(file, offset, data_to_write) ==
    ArrayConcat(ArrayConcat(
        TruncateOrExtendFile(file, offset),
        data_to_write),
        ArraySlice(file, offset + ArrayLen(data_to_write), ArrayLen(file)))
```

> Should any of those utility operators be contributed as an Array module to our CommunityModules?

Yes! I'll look into doing that soon.

IamfromSpace replied:
> your definition for OverlayAt needs to include the portion of the contents after the written data if the contents were not written at the end of the array

Whoops!  Easy to make things simpler if you forget the critical parts :laughing:

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+                []
+                i >= offset /\ i < offset + data_len -> ArrayGet(data_to_write, i - offset)
+                []
+                i >= offset + data_len -> IF i < file_len THEN ArrayGet(file, i) ELSE ArbitrarySymbol])
```
i think it is invariant that the final `ELSE` never matches (unless I've missed something).  The only situation in which is greater than or equal to `offset + data_len` is when the original file extends beyond the total overlap of the new data.

I think it makes sense to avoid the `IF`, as otherwise a cursory glance makes it appear that we can end up `ArbitrarySymbol`s at the end of the file, which isn't the case.

Calvin-L replied:
Yes, great catch! The whole `IF` can be simplified to its then-branch `ArrayGet(file, i)`.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+        //
+        //     Arrays.fill(this.buff, cnt, BuffSz, (byte)0);
+        //
+        // But do not be tempted!  That call incurs a performance penalty WITHOUT OFFERING A STRONGER CONTRACT!  There
```
This is great to call out this optimization (or trap, depending how you're looking at it) and the reasoning behind it.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,175 @@
     private static byte[][] availBuffs = new byte[100][];
     private static int numAvailBuffs = 0;
 
     /* To describe the above fields, we introduce the following
```
Before the spec existed, this is clearly the right place for this comment to exist.  However, given that there is now a formal spec for this module, it seems like it might be appealing to put the bulk of this description into the spec, and then refer the reader to the spec here in this file.

Calvin-L replied:
I am actually going to leave it here, since some of the other comments in this file refer to specific invariants by name, and I think it's helpful to have all that in one file.

But, I will definitely add a note here pointing to the TLA+ formalization.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,175 @@
- * The code was written by Allan Heydon
+ * so that most operations do not require a disk access.
+ *
+ * <p>This class is not thread safe.
```
Since we're specifying and re-examining the properties of this class, it's maybe interesting that it also has slightly weaker guarantees across executions than RAF.  With both RAF and BRAF, should the program crash in the middle of a write to disk, then the value on disk may be corrupted.

However, unique to BRAF, we also lose causality between the state of the file and future side effects.  With RAF, if you do `write()` then `f()`, and you can confirm that `f()` took place, then you also know that the `write()` took place.  With BRAF, it's possible that the `write()` call did not flush before a crash that occurred after `f()`.  So confirming that `f()` took place does not imply that the `write()` call did.

I wouldn't expect this has implications for tlc, but it is interesting to call out as another place where the semantics are weaker.

IamfromSpace replied:
Good point!

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+\* Copyright (c) 2024, Oracle and/or its affiliates.
+
+----------------------- MODULE BufferedRandomAccessFile -----------------------
```
Just wanted to say that I have absolutely found it useful to have a spec of the class for review, and especially as a refinement.

lemmy replied:
This spec could also be referenced at tlaplus/examples

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,175 @@
+     */
+    public BufferedRandomAccessFile(File file, String mode) throws IOException {
+        super(file, mode);
+        synchronized (mu) {
```
The static buffer reuse is interesting.  Just to confirm my understanding: While it would be simpler if each new instance just created its own buffer each time, this appears to be a performance optimization to allow reuse of buffers in new files after old files are closed.  Then, even though this class isn't thread-safe if the same file is modified by two BRAFs, these `synchronized` sections are key to it being thread safe when _distinct_ files all are open with _different_ BRAFs.  Let me know if I've misinterpreted anything!

Did you consider adding this as part of the specification, given a naive danger of reading previous and different file's data?  I'm not sure it's necessary or worth it, as it seems correct, but figured it was worth discussion.

Calvin-L replied:
To be precise:

 - Each BRAF has exclusive access to its internal buffer (`this.buff`).
 - Buffers can be re-used across instances. `availBuffs` is a global "pool" of previously-allocated buffers.
 - The lock `mu` allows BRAFs in different threads to safely interact with the global pool. They do this once on construction (to obtain a buffer) and once on `close()` (to return a buffer to the pool).

The purpose of the shared buffer pool is not clear to me. I suspect it is intended to reduce GC pressure: we can allocate buffers without using `new` and the garbage collector never needs to reclaim them.

However, I _strongly_ suspect that the optimization is irrelevant:
 - I don't believe TLC allocates/closes many instances of `BufferedRandomAccessFile`
 - Even if it does, the JVM's modern GC is _very_ good and may outperform this manual approach
 - And even if I'm wrong about those points, the `open()` syscall probably dominates `BufferedRandomAccessFile` allocation performance anyway

We could look at removing the shared pool in a separate PR.

> Did you consider adding this as part of the specification, given a naive danger of reading previous and different file's data?

I did not. I think the buffer pool problem is orthogonal, and there's a lot of evidence that the current approach works. (I didn't change it.) If I am wrong and it really is great for performance, then we could consider factoring it out into a separate class and specifying its behavior separately.

IamfromSpace replied:
That all tracks!  I think it makes the most sense to leave it as is for now to separate concerns and keep the focus of this PR.  And it makes sense to leave out of the spec given that, like you said, it wasn't changed and its future is dubious.

For discussions sake, as to the possible future of this approach, I agree that its benefit seems suspect.  Also, if there was need to reuse available buffers, the cost of `arraycopy` of all buffers on expansion seems odd.  If we suddenly opened and then closed 1000 files, this is a lot of copying.  I guess it's only the pointers (right?), given it's an array of arrays, but it's still strange to me.

Even if there was a benefit over JVM GC, it seems like atomic interactions with a `Stack` would provide a much more natural interface and should perform better on release.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+    public boolean seeek(long pos) throws IOException {
         // Assert.check(!this.closed);
-        if (pos >= this.hi || pos < this.lo) {
+        boolean didPageRead = false;
```
minor, but with the restructure, there's no need for this intermediate variable.  You can set `this.curr` immediately, and then `return true` at the end of the if condition, and `return false` at the end of the method.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+        // In practice, RandomAccessFile actually moves the pointer if getFilePointer > newLength, regardless of
+        // whether the file was truncated or extended.
+        super.setLength(newLength);
+        if (this.diskPos > newLength) {
```
It may be safer to simply synchronize these unconditionally, given that the documentation already doesn't fully define all behaviors.

Calvin-L replied:
Great suggestion!

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+            this.diskPos = super.getFilePointer();
+        }
+
+        if (this.curr > newLength) {
```
I first thought that maybe this should just always seek to `diskPos`, but I see this is right, because of the subtle cases around end of file.

I think it's worth explaining this case as for why not:
```
If the buffer landed in a way where the end of the file appears in the middle of the buffer,
such as a short file of just 4 bytes.  When we filled the buffer, the diskPos and curr would
both be in the middle of the buffer, at index 4.  Now consider we seek to the beginning of
the file, but then chop one byte off the file.  In this case, `diskPos` will have moved, but
`curr` should not.

         curr and diskPos
         v
[A B C D * * * * ...]

seek(0)

 curr    diskPos
 v       v
[A B C D * * * * ...]

setLength(3)

 curr  diskPos
 v     v
[A B C * * * * * ...]

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+    /**
+     * Write at most "len" bytes from "b" starting at position "off", and return the number of bytes written.
+     */
+    private int writeAtMost(byte[] b, int off, int len) throws IOException {
```
It seems like it may be worth greater optimization here, given that, in theory, that is the motivation for this class.  This implementation makes sense, in that it is quite safe.  If I understand it, we are simply pushing data into the buffer, up to what it can handle, and if we filled the buffer, we use seek to flush the current buffer and move it upwards a single frame, please correct me if I've misunderstood anything!

However, if we write a substantial amount of data, we're a bit wasteful on two fronts.  For one, we copy data into the buffer, which will never be read, when we could directly write the data we have.  For the second, we are going to fill the buffer with data when the frame moves, and do a read from disk for data that we will never read and immediately overwrite.

It seems like we should detect these cases when we're doing a long enough write and bypass the longer route, leaving the buffer in a temporarily invalid state that we know we'll clean up at the very last frame move.  Otherwise, it seems like this class may not be living up to its mission of optimizing file interactions.

Calvin-L replied:
I thought about this as well---but let's defer that to future optimizations. I will add a note about this idea.

FWIW, TLC's workload on this class does not involve large bulk writes. It reads and writes almost entirely single-byte and 4-byte chunks. But, I believe it _does_ alternate between read phases and write phases, so there may be some unnecessary reads during the write phase. Even so, in my microbenchmarks the buffer more than pays for the extra reads.

IamfromSpace replied:
On the topic of future optimizations (not for this PR), it did strike me that making this a ring buffer over an arbitrary position might help avoid cases where the buffer position flaps back and forth between two frames, causing a series of reads and flushes.  It also might better work with an L2 cache by keeping buffer positions more stable in memory for longer.

Though, again, I’d be curious your estimation of how such changes might actually impact the TLC specific workload.

IamfromSpace commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,384 @@
+
+\* Inv2 is a bit special.  Most methods restore it just before they return.  It
+\* is generally restored by calling `seeek(curr)`.  But, that behavior is
+\* difficult to model in straight TLA+ because each method may modify variables
```
I liked how Finn and co. did a mix of TLA+ and PlusCal in the [Inconsistency in Cosmos DB paper](https://arxiv.org/pdf/2210.13661.pdf) to play to the two's strengths and get clear multi-step behaviors.  It might be worth it to look at something like that for when variables need multiple changes, and then the invariants would only be expected to hold when the program counter is in safe state.

Calvin-L replied:
I actually started with PlusCal, but even though it makes modeling easier it adds a lot of specification overhead. Specifically, verifying that the output of `read()` is correct might require an extra `read_output` variable. It gets a little ugly if `read()` is not a single action.

But, definitely something to keep in mind if this class becomes any more complex.

Calvin-L commented:
Thanks a ton @IamfromSpace for the feedback!

I think I've addressed all the major comments so far.

IamfromSpace commented:
These updates look great!  For the sake of thoroughness I still have a few comments coming through on tests, but nothing of real concern.  Just wrapped up the unit tests, but it's taking me a bit longer to digest the fuzzing file :)

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileTest.java:
```diff
@@ -80,8 +77,8 @@ public void testWriteSeekNoLength() throws IOException {
 			return;
-		} catch (Exception e) {
-			fail(e.getMessage());
+//		} catch (Exception e) {
```
Was this never necessary?  I haven't been able to figure out why this was ever here in the first place :laughing:

Calvin-L replied:
It was not. 😆 It does exactly one thing, which is to hide the stack trace from the thrown exception `e`. That one thing is bad, so I've removed these catch blocks.

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileTest.java:
```diff
@@ -104,6 +101,7 @@ public void testRead() throws IOException {
 	}
 	
 	@Test
+	@Ignore("BufferedRandomAccessFile does not (and never did) guarantee zero-fill when extended")
```
nit: I think it makes more sense to just delete this.

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileTest.java:
```diff
@@ -170,4 +168,155 @@ public void testInvalidateBufferedData() throws IOException {
+	}
+
+	@Test
+	public void regressionTest01() throws IOException {
```
nit: I personally like giving these clearer names.  It's easier to read them if you have that context of what they intend to validate (or what you broke if they fail!).

Calvin-L replied:
There's a lot to say about these `regressionTest*` methods that should be written in a comment.

TL;DR: These were generated by a computer, so they don't look particularly sensible, nor can I really assign them sensible names. I don't plan to change them (see below), but I am open to adding additional focused handwritten tests inspired by them.

----

The `regressionTest*` methods are all randomly-generated failures discovered by the fuzzer at some point while I was working on this PR. I copied them unchanged from the fuzzer output (although in a few cases I added a comment about what should happen for a particular test case).

The fact that these cases failed during development indicates they are testing _something_ that is tricky to get right in the implementation, and so I think it's worth having them here to make sure they are tested regularly.

I don't particularly want to change them because:
 - I am not confident I can minimize them reliably; the failures happened on intermediate versions of `BufferedRandomAccessFile` that are now lost. If I make a change and the test still passes, is it _really_ testing the same thing it was before?

And I don't want to remove them either:
 - Some of these came from an earlier version of the fuzzer (i.e. the current fuzzer doesn't cover them exactly as written here).
 - Small changes to the fuzzer generally result in large changes to explored traces (i.e. future versions of the fuzzer may not explore these specific cases).

IamfromSpace replied:
Sounds good, like a i said, no deal breakers for me at this point :)  I think the test updates you made are already above and beyond!

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileTest.java:
```diff
@@ -170,4 +168,155 @@ public void testInvalidateBufferedData() throws IOException {
+		try (final java.io.RandomAccessFile raf = new BufferedRandomAccessFile(tmpFile, "rw")) {
+			raf.seek(9037);
+			raf.setLength(8311);
+			byte[] b1 = { 114, -128, 118, -15, -71, -43, -70, -27, 122, -33, -8, 80, -75, -7, -123, -88, -109, 4, 124, -69, 73, -86, 21, -63, 84, -13, 106, 81, 64, 69, -109, -56, -87, -48, 66, -98, -83, 44, 111, -77, 78, 17, -123, 115, -18, -55, -42, -75, 73, -73, -23, -125, -113, -39, 85, -39, -119, -16, 76, -106, -47, -72, 113, 25, 78, -90, 35, -2, -109, 1, 63, 43, 14, 44, -112, 8, -15, 15, 18, -82, -20 ...
```
nit: this seems a bit overkill.  Can't this just be a `raf.write(0)` and then assert that length is 8312?

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileTest.java:
```diff
@@ -170,4 +168,155 @@ public void testInvalidateBufferedData() throws IOException {
+		try (final java.io.RandomAccessFile raf = new BufferedRandomAccessFile(tmpFile, "rw")) {
+			raf.seek(9247);
+			raf.setLength(1175);
+			byte[] b1 = { 28, -61, -90, 105, -37, 115, 76, 15, -2, -82, -61, 66, -77, 73, -96, 119, 102, 8, 126, 40, -6, -80, 35, 112, 6, -9, -128, 72, 37, -20, 29, 119, -116, 110, 29, -83, -26, -63, 28, 79, -55, 75, 68, -44, -77, 13, 22, -58, 13, 5, 76, 88, 124, 104, -19, -87, -9, 121, -43, -83, 44, 84, 89, 39, 61, 60, -113, -46, -26, 56, 113, -121, 60, 90, -46, -70, 37, -95, 10, 18, 89, 21, 103, -77, 30 ...
```
nit: It seems like we can create this case more easily.

Doesn't the following trigger the original `AssertionError`?

```
raf.seek(9247);
raf.setLength(71);
raf.seek(1675);
raf.write(87);
```

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java:
```diff
@@ -0,0 +1,553 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class BufferedRandomAccessFileFuzzTest {
```
This is great, I'm a big fan of fuzz testing.  There's a lot of code here, but I think I'm maybe I've become spoiled in the Haskell and QuickCheck world :)

My main recommendation is that we call out the overall strategy here to present a mental framework for anyone who needs to work in here in the future.  The main thing that I was slow to put together was that (and how) invalid lists of operations were pruned.  But in retrospect, the approach is quite obvious, so I think a short blurb would go a long way.

Something like: "Fuzzing here generates many random legal lists of file operations and then validates that executing those operations has the same outcome, whether applied to a `RandomAccessFile` or a `BufferedRandomAccessFile`.  When randomly generating operation lists, we simulate these lists on an in memory representation of the file (`AbstractFileState`) which has helpers to indicate if an operation that would result in undefined behavior.  Such illegal operations are then pruned before checking."

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java:
```diff
@@ -0,0 +1,553 @@
+        return ops;
+    }
+
+    private void generateRandomOpsAndCheck(Random rng, int nOps) throws IOException {
```
This function appears to be redundant.  Even if it had some value in debugging (taking a guess at its purpose), I still think it makes more sense to remove it given that it has otherwise duplicated code.

Calvin-L replied:
Good catch; this is leftover from some intermediate version and should be removed.

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java:
```diff
@@ -0,0 +1,553 @@
+        }
+    }
+
+    private RunResult checkOps(java.util.List<FileOperation<?>> ops) throws IOException {
```
I might recommend calling this `checkLegalOps`.  I think `check` and `legal` make as much sense as any names, but I did have a moment of confusion between the purpose of the two.  This name might make clear that the function's domain is only legal operations, and we should therefore always expect that we're going to have to check their legality first.

Calvin-L replied:
Now that you mention it, I think "legal" should be "well-defined". I used the two terms interchangeably, but I think "well-defined" is a better name for it.

IamfromSpace replied:
Nice, yeah, I think this is a much clearer change!

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java:
```diff
@@ -0,0 +1,553 @@
+
+    private RunResult checkOps(java.util.List<FileOperation<?>> ops) throws IOException {
+
+		final File tmpFile1 = File.createTempFile("tmp", ".bin");
```
nit: spacing appears to be accidental

IamfromSpace commented on tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java:
```diff
@@ -0,0 +1,553 @@
+        return new RunResult(ops, ops.size(), null, null);
+    }
+
+    private boolean legal(java.util.List<FileOperation<?>> ops) {
```
minor, but it struck me that your construction of `FileOperation` here is pretty clever, to where it seems like like could instead be the `simulateIfLegal` method on `ListOfOperations` class that implements `FileOperation`.  Seems like it has a valid `execute` and `toString`.

Calvin-L replied:
Indeed! But there is a fair amount of code that cares about the list structure, so I don't think there's much benefit in implementing `SequencedOperations implements FileOperation`, even if it would be very elegant. :)

IamfromSpace reviewed (approved):
Alrighty, that's 100% everything from me :)  I think there's some test comments that are worth a look over, but there's certainly no deal-breakers in there, so I've got the approval on there now.  Not sure if the repo is configured to reset on changes or not, but if so, I'll keep it active for any commits that come in.

Nice work on this!

lemmy commented:
I won't have time to review this PR anytime soon and I don't want to be a bottleneck. Given @IamfromSpace's review, I feel confident about this PR. @Calvin-L, feel free to merge or consider inviting additional reviewers at the TLA+ Community Meeting in two weeks. Reviewing the spec offers a great opportunity for those unfamiliar with or not fond of Java to contribute to the TLA+ tools.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,190 @@
+     */
+    private long mark;
+
+    /** Protects {@link #availBuffs} and {@link #numAvailBuffs}. */
```
`mu` also seems to protect `buff`.

Calvin-L replied:
Not quite; `availBuffs` and `numAvailBuffs` implement a shared [free list](https://en.wikipedia.org/wiki/Free_list) for allocating buffers. Once this class has allocated a buffer by removing one from `availBuffs`, it does not need to lock anything to access its privately allocated `buff`.

More details in this comment --> https://github.com/tlaplus/tlaplus/pull/907#discussion_r1573045418

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+        // But do not be tempted!  That call incurs a performance penalty WITHOUT OFFERING A STRONGER CONTRACT!  There
+        // are still cases where we rely on "super" to extend the file---meaning we inherit the weakest possible
+        // contract from RandomAccessFile.  In particular, this happens when extending the file using setLength() or
+        // when the client makes a giant forward seek followed by a write.
```
This comment could imply a possible completeness bug in `DiskFPSet` s.t. the arbitrary data matches a fingerprint that is probed.  This particular bug would vanish if the content of the buffer was zeroed.

Calvin-L replied:
I did a small audit of `DiskFPSet` and its subclasses when I was working on this and determined that they don't actually read arbitrary data. After `setLength()` they make an effort to fill the entire allocated file with data. But, it's quite possible I missed something.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
-    /* overrides RandomAccessFile.seek(long) */
+    @Override
     public void seek(long pos) throws IOException {
+        this.seeek(pos); // ignore boolean result
```
Can we remove `seeek` given that `seek` delegates to `seeek`?

lemmy replied:
Let me clarify: Can we replace `seek` with `seeek`? :-)

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,446 @@
+\* is generally restored by calling `seeek(curr)`.  But, that behavior is
+\* difficult to model in straight TLA+ because each method may modify variables
+\* multiple times.  So instead, this spec treats Inv2 as a precondition for the
+\* methods and verifies that it is indeed restored by `Seek(pos)`, which should
```
`Inv2CanAlwaysBeRestored` suggests that this should mention `FlushBuffer` in addition to `Seek`.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,446 @@
+
+\* Inv2 is a precondition for many actions; it should always be possible to
+\* restore Inv2.
+Inv2CanAlwaysBeRestored ==
```
I translated the prose comment into the following property:

```tla
Inv2CanAlwaysBeRestoredPart2 ==
    [][IF dirty THEN FlushBuffer => Inv2' ELSE Seek(curr) => Inv2']_vars
```

, which TLC says is violated by the following trace:

```tla
Finished computing initial states: 1 distinct state generated at 2024-06-03 15:53:42.
Action property Inv2CanAlwaysBeRestored is violated.
The behavior up to this point is:
1: <Initial predicate>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = FALSE
/\ length = 0
/\ curr = 0
/\ lo = 0
/\ buff = [elems |-> <<ArbitrarySymbol, ArbitrarySymbol>>]
/\ diskPos = 0
/\ file_content = [elems |-> <<>>]
/\ file_pointer = 0
/\ abstract_contents = [elems |-> <<>>]
2: <WriteAtMost([elems |-> <<A, A>>]) line 193, col 5 to line 202, col 60 of module BufferedRandomAccessFile>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = TRUE
/\ length = 2
/\ curr = 2
/\ lo = 0
/\ buff = [elems |-> <<A, A>>]
/\ diskPos = 0
/\ file_content = [elems |-> <<>>]
/\ file_pointer = 0
/\ abstract_contents = [elems |-> <<A, A>>]
3: <FlushBuffer line 109, col 5 to line 119, col 43 of module BufferedRandomAccessFile>
/\ BuffSz = 2
/\ MaxOffset = 4
/\ dirty = FALSE
/\ length = 2
/\ curr = 2
/\ lo = 0
/\ buff = [elems |-> <<A, A>>]
/\ diskPos = 2
/\ file_content = [elems |-> <<A, A>>]
/\ file_pointer = 2
/\ abstract_contents = [elems |-> <<A, A>>]
```

Revisiting https://github.com/tlaplus/tlaplus/pull/907/files#r1625145343, this property, perhaps, too strong.  This hypothesis is supported because the following holds:

```tla
Inv2CanAlwaysBeRestoredPart2 ==
    [][Seek(curr) => Inv2']_vars
```

Calvin-L replied:
This is a fantastic point. I will look into strengthening this property on Monday. Perhaps it can be fully specified using `\cdot` to compose `FillBuffer` and `Seek(curr)`.

lemmy replied:
+1, I also forgot to mention that I would advise against using `\cdot` (gut feeling).

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,446 @@
+WriteToFile(file, offset, data_to_write) ==
+    LET file_len == ArrayLen(file) IN
+    LET data_len == ArrayLen(data_to_write) IN
+    LET length == Max(file_len, offset + data_len) IN
```
Nitpick and you probably know this, but I prefer a single LET/IN:

```tla
LET file_len == ArrayLen(file)
    data_len == ArrayLen(data_to_write)
    length == Max(file_len, offset + data_len) IN
```

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,446 @@
+    /\ numReadableWithoutSeeking >= 0
+    /\ LET numToRead == Min(ArrayLen(data), numReadableWithoutSeeking) IN
+       LET buffOff == curr - lo IN
+        /\ data = ArraySlice(buff, buffOff, buffOff + numToRead)
```
Neither `Read` nor `Read1` are enabled if the buffer contains `ArbitrarySymbol`, i.e., the following invariant holds:

```tla
DebugReadsWithArbitrarySymbol ==
    \E i \in 1..Len(buff.elems): buff.elems[i] = ArbitrarySymbol =>
        /\ \A data \in UNION{ Array(Symbols, len): len \in 1..MaxOffset }:
                ~ENABLED Read(data)
        /\ \A symbol \in Symbols:
                ~ENABLED Read1(symbol) 
```

Why this restriction, given that `BufferedRandomAccessFile.java` does not impose such a restriction?

Calvin-L replied:
You are right; I believe that restriction can and should be relaxed.

Calvin-L replied:
@lemmy There is an associativity problem in your definition; I think it should be
```tla
DebugReadsWithArbitrarySymbol ==
    (\E i \in 1..Len(buff.elems): buff.elems[i] = ArbitrarySymbol) => \* NOTE parens around \E expression
        /\ ...
```

Using that modified definition shows that `Read` actually is enabled when the read does not cover `ArbitrarySymbol`, even if there are `ArbitrarySymbol`s in the buffer.

Even so, I do think it makes sense to allow reads of `ArbitrarySymbol` in the next-relation.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla:
```diff
@@ -0,0 +1,446 @@
+BufferedIndexes == lo .. (Min(lo + BuffSz, length) - 1)
+
+Inv1 ==
+    /\ diskPos = file_pointer
```
This differs from `V1` in BufferedRandomAccessFile.java.

FedericoPonzi commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
-                System.arraycopy(availBuffs, 0, newBuffs, 0, numAvailBuffs);
-                availBuffs = newBuffs;
+        try {
+            // Assert.check(!this.closed);
```
currently, the implementation allows read after close. We close the underlying file, give back the buffer to the pool but the buffer is not nullified, and subsequent reads that don't access the file will succeed.

Sample test adopted from the testRead (see last three rows)
```
	@Test
	public void testRead() throws IOException {
		final File tmpFile = File.createTempFile("BufferedRandomAccessFileTest_testRead", ".bin");
		tmpFile.deleteOnExit();
		final java.io.RandomAccessFile raf = new BufferedRandomAccessFile(tmpFile, "rw");
		for (long i = 0L; i < BufferedRandomAccessFile.BuffSz / 8; i++) {
			raf.writeLong(i);
		}

		raf.seek(0);
		for (long i = 0l; i < BufferedRandomAccessFile.BuffSz / 8; i++) {
			assertEquals(i, raf.readLong());
		}
                raf.seek(0); //seek back to zero
		raf.close(); // closed
                assertEquals(0, raf.readLong()); // still works
	}
```
This could be a problem in case the buffer is reassigned to a different BRAF. To reduce the state, it would be sufficient to call "isClosed" on the underlinying RAF. Because there is no isClosed method, getFilePointer could be used for its side effect of failing if the underlying object is closed.

Calvin-L replied:
I have mixed feelings about this.

> the implementation allows read after close

Yes and no. The implementation does not throw an exception if you do a read after close, so in that sense it "allows" read after close. But, the implementation also doesn't specify behavior if you read after close, so in that sense it "disallows" read after close. In other words, _clients_ are responsible for ensuring that they do not interact with closed files.

In this case, I would peronsally rather push responsibility to the callers---but it's true that the official [RandomAccessFile contract](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/RandomAccessFile.html) _does_ require `IOException` to be thrown for _some_ read and write methods if the file is closed (but weirdly, not all of them). In your specific example there is no specified behavior, and therefore the current behavior ("return an arbitrary long without throwing an exception") is legal. But if you replace the final `readLong()` with `read(byte[])`, it is indeed a demonstration of a bug, since that method requires an exception if the file is closed.

Also, because you called this out I noticed that this `close()` method is not idempotent, and [it must be since this class implements `Closeable`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/AutoCloseable.html). A follow-up call will return `buff` to the `availBuffs` pool a second time, which could be catastrophic.

In any case, I think there is a strong case for reinstating the commented-out `closed` field, which would make it very easy to fix these problems. I will look into doing this on Monday.

Calvin-L replied:
Aside: it is also irksome to me that the official contract for `RandomAccessFile` is to throw `IOException` if the file is closed, when [`IllegalStateException`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/IllegalStateException.html) already exists and was designed for _exactly this purpose_:

> Signals that a method has been invoked at an illegal or inappropriate time.

But, to my surprise, `RandomAccessFile` is _older_ than `IllegalStateException`, so I suppose it makes sense.

Calvin-L replied:
@FedericoPonzi @lemmy On second thought, I do not plan to address this comment in this PR. There is no _new_ bug here, so I think it makes sense to address this in a separate PR.

FedericoPonzi reviewed (commented):
Very good job! I haven't got to the end yet, I've found a read after close issue and add some nits.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+    // NOTE: a call to seeek(this.curr) suffices to restore invariant V2 after this.curr changes.
+    //TODO come up with better name for seeek %)
+    public boolean seeek(long pos) throws IOException {
         // Assert.check(!this.closed);
```
Generally yes, but---

I plan to keep this (and some other dead code you called out) for two reasons:
 1. `closed` is mentioned in the class invariants; it is (what I would call) a "ghost variable" that is part of the specification but not part of the implementation. So, its presence has a little bit of value.
 2. Keeping this field reduces the already-quite-large size of this PR diff. We can consider removing it (and other dead code) in a separate maintenance-only PR.

lemmy replied:
@FedericoPonzi In addition to what @Calvin-L said, we generally try to keep commits as small as possible by, for example, limiting them to a single concern. This makes future bisecting simpler should it become necessary. For instance, if this PR introduces a regression, diagnosing the issue will be much easier if the changes are minimal.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+     * @return true iff data was read from disk
+     */
+    // NOTE: a call to seeek(this.curr) suffices to restore invariant V2 after this.curr changes.
+    //TODO come up with better name for seeek %)
```
This TODO is still relevant; a future maintenance-only PR could rename `seeek` to something more meaningful. :)

This TODO was moved or re-indented, which is why it appears in the diff. It might be easier to review this while hiding whitespace-only changes:
<img width="674" alt="Screenshot 2024-06-07 at 1 02 05 PM" src="https://github.com/tlaplus/tlaplus/assets/1393639/c610c6a8-7b85-4fea-95a3-6cf2f8772cb4">

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+
+    @Override
     public long length() throws IOException {
         // Assert.check(!this.closed);
```
See note above about `closed`.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,190 @@
     private boolean dirty;  // true iff unflushed bytes exist
+
     // SZ Feb 24, 2009: never read
     // private boolean closed; // true iff the file is closed
```
See note above about `closed`.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,190 @@
-    {
+    @Deprecated
+    public BufferedRandomAccessFile(String name, String mode) throws IOException {
         // Simon Z. replaced the original:
         //    super(name, mode);
         //    this.init();
         // with this.
```
See [note above about `closed`](https://github.com/tlaplus/tlaplus/pull/907#discussion_r1631637066).

Er, well, this one doesn't refer to `closed` specifically, but I am retaining this dead code to keep the size of this PR down.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,190 @@
+     */
+    private void init() throws IOException {
+        this.dirty = false;
+        // SZ Feb 24, 2009: never read locally
+        // this.closed = false;
```
See [note above about `closed`](https://github.com/tlaplus/tlaplus/pull/907#discussion_r1631637066).

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+     * @return true iff any disk writes were performed
+     */
     private boolean flushBuffer() throws IOException {
         if (this.dirty) {
```
Well, [the original code was structured this way too](https://github.com/tlaplus/tlaplus/blob/2bfa142d595e54487b74a793cfb5578f091f2b82/tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java#L202), so this change wouldn't be a "revert" per se. In this case I would rather preserve the original structure as much as possible to keep this PR diff smaller and more focused.

In general I agree with you; an early return would be a clearer and more canonical way to structure the method, and that's how I would write it if I were writing this method from scratch. But in this case the existing method follows the "single return principle" (or perhaps the "returns should all be close to each other principle"), which I think is a perfectly acceptable alternative to early returns when there aren't too many nested ifs.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -5,149 +5,190 @@
+    // longer with larger buffer sizes.
+    static final int LogBuffSz = 13; // 8K buffer
     public static final int BuffSz = (1 << LogBuffSz);
     static final int BuffMask = ~(BuffSz - 1);
```
Tricky, but you are right! I will make this change on Monday and also add a comment about what this field is computing. It is not at all obvious why `-BuffSz` is a meaningful thing to compute (or for that matter why `~(BuffSz - 1)` is meaningful either).

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+        try {
+            // Assert.check(!this.closed);
+            this.flush();
+            // SZ Feb 24, 2009: never read locally
+            // this.closed = true;
```
See [note above about `closed`](https://github.com/tlaplus/tlaplus/pull/907#discussion_r1631637066).

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+            // SZ Feb 24, 2009: never read locally
+            // this.closed = true;
+            synchronized (mu) {
+                // grow "availBuffs" array if necessary
```
Yes, or removed entirely (which would be [my personal preference](https://github.com/tlaplus/tlaplus/pull/907#discussion_r1573045418)). I retained this code to keep this PR more focused; [this code already existed](https://github.com/tlaplus/tlaplus/blob/2bfa142d595e54487b74a793cfb5578f091f2b82/tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java#L181).

A future PR might remove or alter this for improved performance or maintainability.

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
-      return new BigInteger(b);
+        // If we read past the buffer, restore our internal invariants
+        if (this.curr >= this.lo + BuffSz) {
+            seeek(this.curr);
```
I am inclined to factor this out into a helper method with a name like `restoreInv2`, since an identical check and `seeek` call appears other places as well. I will make that change soon. (When I do, I will probably keep the call to `seeek` and not `seek`, since it avoids an extra call stack entry.)

Calvin-L commented on tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java:
```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
+                if (this.diskPos != this.lo) {
+                    super.seek(this.lo);
+                }
+                super.write(this.buff, 0, len);
```
It is weird, but [`write` takes an int as its third argument](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/RandomAccessFile.html#write(byte%5B%5D,int,int)). That is because Java arrays can only contain an `int`'s worth of entries, and the `len` argument specifies how many entries to read from the `buff` array.

In this case `len` is guaranteed to fit in an `int`, since it can't be larger than `BuffSz` and `BuffSz` is an `int`.

Calvin-L commented:
Thank you for the feedback, everyone! I believe I've addressed all the comments so far. There are currently four follow-up items to look into after merging:
 - Add an `Arrays` module to the Community Modules
 - Reference this spec from tlaplus/examples once it is merged
 - Verify that `DiskFPSet` and friends are not reading undefined data from the file after extending it with `setLength`
 - Make `close()` idempotent and make reads and writes throw `IOException` for closed files
